### PR TITLE
Feature: integrate dice-group/metall-ffi

### PR DESCRIPTION
**Pull request by liss-h**
Integrates [dice-group/metall-ffi](https://github.com/dice-group/metall-ffi)

### CODE REVIEW

File: libs/ffi/src/dice/ffi/metall.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct metall_manager metall_manager;
+
+/**
+ * @brief Attempts to open the metall datastore at path
+ * @param path path to datastore
+ * @return true on success, false on failure. On failure, sets errno to one of the following values:
```
Comment (bigerl) [inline libs/ffi/src/dice/ffi/metall.h]: return doc is wrong

File: libs/ffi/src/dice/ffi/metall.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ *      - ENOTRECOVERABLE if the given metall datastore is inconsistent
+ */
+metall_manager *metall_open(char const *path);
+
+/**
+ * @brief Attempts to open the metall datastore at path in read only mode
+ * @param path path to datastore
+ * @return true on success, false on failure. On failure, sets errno to one of the following values:
```
Comment (bigerl) [inline libs/ffi/src/dice/ffi/metall.h]: return doc is wrong

File: libs/ffi/src/dice/ffi/metall.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ *      - ENOTRECOVERABLE if the given metall datastore is inconsistent
+ */
+metall_manager *metall_open_read_only(char const *path);
+
+/**
+ * @brief Attempts to create a metall datastore at path
+ * @param path path at which to create a datastore
+ * @return true on success, false on failure. On failure, sets errno to one of the following values:
```
Comment (bigerl) [inline libs/ffi/src/dice/ffi/metall.h]: return doc is wrong

File: libs/ffi/src/dice/ffi/metall.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+extern "C" {
+#endif
+
+typedef struct metall_manager metall_manager;
+
+/**
+ * @brief Attempts to open the metall datastore at path
+ * @param path path to datastore
+ * @return true on success, false on failure. On failure, sets errno to one of the following values:
+ *      - ENOTRECOVERABLE if the given metall datastore is inconsistent
```
Comment (bigerl) [inline libs/ffi/src/dice/ffi/metall.h]:
```suggestion
 *      - ENOTRECOVERABLE if the given metall datastore is inconsistent or already in use
```

File: libs/ffi/src/dice/ffi/metall.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ */
+metall_manager *metall_open(char const *path);
+
+/**
+ * @brief Attempts to open the metall datastore at path in read only mode
+ * @param path path to datastore
+ * @return true on success, false on failure. On failure, sets errno to one of the following values:
+ *      - ENOTRECOVERABLE if the given metall datastore is inconsistent
```
Comment (bigerl) [inline libs/ffi/src/dice/ffi/metall.h]:
```suggestion
 *      - ENOTRECOVERABLE if the given metall datastore is inconsistent or already in use
```

File: libs/ffi/src/dice/ffi/metall.cpp [hunk 6]
```diff
@@ -1,60 +1,107 @@
...
+
+  return reinterpret_cast<metall_manager *>(manager);
 }
 
-void metall_close() { delete g_manager; }
+bool metall_snapshot(metall_manager *manager, char const *dst_path) {
+  return reinterpret_cast<metall_manager_t *>(manager)->snapshot(dst_path);
+}
 
-void metall_flush() { g_manager->flush(); }
```
Comment (bigerl) [inline libs/ffi/src/dice/ffi/metall.cpp:31]: Is there a particular reason why the flush function was dropped?

**Review by bigerl — CHANGES_REQUESTED** [review]
Thx. Looks good. Only some minor things left.

File: libs/ffi/src/dice/ffi/metall.cpp (hunk 6 above)
Comment (liss-h) [reply libs/ffi/src/dice/ffi/metall.cpp:31]: No, just never existed in our version of the FFI. Will add it again

**Review by bigerl — APPROVED** [review]

**Pull request merged**